feat(docs-infra): created new widget for events page

ajitsinghkaler opened the pull request:
Earlier data in events page was hardcoded and data have to be moved manually in the table now events will automatically move between current and past events in this implementation

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Data has to be manually changed in events

Issue Number: N/A


## What is the new behavior?
Data will automatically be changed in presented and upcoming events

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ajitsinghkaler commented:
@petebacondarwin @StephenFluin @gkalpak  if we can finalize the way in which the new events widget will be implemented after that I will add tests

mary-poppins commented:
You can preview bf62aab at https://pr36517-bf62aab.ngbuilds.io/.
You can preview 3be5189 at https://pr36517-3be5189.ngbuilds.io/.

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,170 @@
+[
+  {
+    "endDate": "2019-09-20",
+    "address": "London, UK",
```
I'd vote for calling this property "location" (or "where").
(It's not a proper address.)

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,170 @@
+  {
+    "endDate": "2019-09-20",
+    "address": "London, UK",
+    "title": "AngularConnect",
```
This is essentially used as a tooltip. I suggest we rename it to "tooltip" (because "title" is confusing).

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+
+import { EventsService } from './events.service';
+
+export interface Event {
+  endDate: string;      // "2020-05-27",
+  address: string;      // "Oslo, Norway",
+  title: string;        // "ngVikings",
+  linkUrl: string;      // "https://ngvikings.org/",
+  name: string;          // "ngVikings",
+  info: string;          // "May 25-26 conference, 27 workshops, 2020"
+}
```
In general, I suggest the following format for each event:

```ts
type date = String; // of the format `YYYY-MM-DD`.

interface Event {
  name: string;
  location: string;
  linkUrl: string;
  tooltip?: string;  // Optional - defaults to `name`.

  date: {
    start: date;
    end: date;
  };

  workshopsDate?: {
    start: date;
    end: date;
  };
}
// ...where `date` is a string of the format `YYYY-MM-DD`.
```

If we find in the future that this type is too restrictive, we can easily convert it to something more flexible, but I think it would suffice for our current needs.

NOTE: We can compute where to show each event based on the dates in `date` and `workshopsDate`.

WDYT?

ajitsinghkaler replied:
did it but only on few events seems a bit cluttered

gkalpak replied:
_Continuing the discussion in https://github.com/angular/angular/pull/36517#discussion_r407496303..._

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+  selector: 'aio-events',
+  templateUrl: 'events.component.html'
+})
+export class EventComponent implements OnInit {
```
```suggestion
export class EventsComponent implements OnInit {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+})
+export class EventComponent implements OnInit {
+
+  prevEvents: Event[];
```
Nit: I think "past events" is more clear:

```suggestion
  pastEvents: Event[];
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+  prevEvents: Event[];
+  upcomingEvents: Event[];
+
+  constructor( private eventsService: EventsService ) { }
```
Nit:

```suggestion
  constructor(private eventsService: EventsService) { }
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+  constructor( private eventsService: EventsService ) { }
+
+  ngOnInit() {
+    this.eventsService.events.subscribe(( events: Event[] ) => {
```
Nit: The type shouldn't be needed:

```suggestion
    this.eventsService.events.subscribe(events => {
```

...but if you insist on having it, please remove the surrounding whitespace:

```suggestion
    this.eventsService.events.subscribe((events: Event[]) => {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+
+  ngOnInit() {
+    this.eventsService.events.subscribe(( events: Event[] ) => {
+      this.prevEvents = events.filter( event => {
```
```suggestion
      this.prevEvents = events.filter(event => {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+  ngOnInit() {
+    this.eventsService.events.subscribe(( events: Event[] ) => {
+      this.prevEvents = events.filter( event => {
+        return new Date(event.endDate).valueOf() < Date.now();
```
`getTime()` is more idiomatic:

```suggestion
        return new Date(event.endDate).getTime() < Date.now();
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+        return new Date(event.endDate).valueOf() < Date.now();
+      });
+
+      this.upcomingEvents = events.filter( event => {
```
```suggestion
      this.upcomingEvents = events.filter(event => {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,36 @@
+      });
+
+      this.upcomingEvents = events.filter( event => {
+        return new Date(event.endDate).valueOf() > Date.now();
```
So, for a very brief moment in time an event won't be listed in any table :grin:

```suggestion
        return new Date(event.endDate).valueOf() >= Date.now();
```

Actually, I think the logic is not correct for determining where to show an event. When using `new Date('YYYY-MM-DD')` it implies `00:00` as time. I think we should show the event as "upcoming" during the last day and only move it to "past events" after that.

So, insted of comparing against `Date.now()`, we should be comparing against `Date.now() - (/* One day */ 24 * 60 * 60 * 1000)`.

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,170 @@
+    "info": "January 30, 2018"
+  },
+  {
+    "endDate": "2020-05-27",
```
Please, sort this file by event date 🙏

gkalpak commented on aio/src/app/custom-elements/events/events.service.ts:
```diff
@@ -0,0 +1,29 @@
+  private getEvents() {
+    return this.http.get<any>(resourcesPath).pipe(
+      map(data => data.sort(compareEvents))
+    );
```
We should catch errors too. (See [announcement-bar.component.ts][1] for an example.)

[1]: https://github.com/angular/angular/blob/3be5189dd0a42e8257c4b4ae20f774a6cdcc7848/aio/src/app/custom-elements/announcement-bar/announcement-bar.component.ts

gkalpak commented on aio/src/app/custom-elements/events/events.service.ts:
```diff
@@ -0,0 +1,29 @@
+}
+
+function compareEvents( l: Event, r: Event ) {
+  return (new Date(l.endDate) < new Date(r.endDate) ? 1 : -1);
```
Since dates are in `YYYY-MM-DD` format, the default lexicographic sorting should be enough.
(It will become problematic in AD 10000, but we have time until then :grin:)

gkalpak reviewed (changes requested):
I've wanted this for a long time, thx, @ajitsinghkaler :tada:
I've left a first round of comments.

mary-poppins commented:
You can preview 5332c91 at https://pr36517-5332c91.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes but made the event data changes only on few events once we confirm it we can migrate the whole data

mary-poppins commented:
You can preview 171f7c0 at https://pr36517-171f7c0.ngbuilds.io/.

gkalpak commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -0,0 +1,49 @@
+      <td><a href="{{event.linkUrl}}" title={{event.tooltip}}>{{event.name}}</a></td>
+      <td>{{event.location}}</td>
+      <td>
+        <div>
+          {{event.date.start | date:'LLLL d, y'}} - {{event.date.end | date:'LLLL d, y'}}
+        </div>
+        <div *ngIf="event.workshopsDate">
+          Workshops: {{event.workshopsDate.start | date:'LLLL d, y'}} - {{event.workshopsDate.end | date:'LLLL d, y'}}
+        </div>  
```
_Continuing the discussion from https://github.com/angular/angular/pull/36517#discussion_r407178203..._

This is indeed quite verbose. The benefit of having the data in a structured format is that we can display them however we want and still get a consistent representation for all events. Previously, each event used arbitrary formatting; some examples:
- `January 1-3, 2020`
- `Jan 1-3, 2020`
- `January 31 - February 2, 2020`
- `Jan 1st workshops, 2nd conference, 2020`
- `January 1-2 conference, 3 workshops, 2020`
- `January 1 (workshops), 2-3 (conference), 2020`

Now, we have the raw data for each event and we can apply consistent formatting. Below are some usecases we need to take into account (with my suggested formatting in parenthesis):

- One-day conference.
  _(Suggestion: `January 1, 2020`)_
- Multi-day conference - same month.
  _(Suggestion: `January 1-3, 2020`)_
- Multi-day conference - different month.
  _(Suggestion: `January 31 - February 2, 2020`)_
- Conference with multi-day workshops before main event - same month.
  _(Suggestion: `January 1-2 (workshops), 3-4 (conference), 2020`)_
- Conference with one-day workshops before main event - same month.
  _(Suggestion: `January 1 (workshops), 2-3 (conference), 2020`)_
- Conference with one-day workshops after main event - same month.
  _(Suggestion: `January 1-2 (conference), 3 (workshops), 2020`)_
- Conference with multi-day workshops before main event - different month.
  _(Suggestion: `January 30-31 (workshops), February 1-2 (conference), 2020`)_
- Conference with one-day workshops before main event - different month.
  _(Suggestion: `January 31 (workshops), February 1-2 (conference), 2020`)_
- Conference with one-day workshops after main event - different month.
  _(Suggestion: `January 30-31 (conference), February 1 (workshops), 2020`)_

ajitsinghkaler replied:
What are your thoughts should I use ngSwitch case or what shall I use

gkalpak replied:
There is some non-trivial logic involved, so this should be handled in the component imo. I.e. have a helper that takes `event.date` and `event.workshopsDate` and return a string.

gkalpak reviewed (changes requested):
Thx for making the changes, @ajitsinghkaler 🚀
Here's a second round of comments :grin:

petebacondarwin commented:
Happy to let @gkalpak take the lead on reviewing this. If you want extra input from me, please re-add me as a reviewer when it is ready.

mary-poppins commented:
You can preview b473d7b at https://pr36517-b473d7b.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes please have a look

mary-poppins commented:
You can preview 8d3bd94 at https://pr36517-8d3bd94.ngbuilds.io/.

ajitsinghkaler commented:
I will make linitng changes later

mary-poppins commented:
You can preview 53c2708 at https://pr36517-53c2708.ngbuilds.io/.

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    });
+  }
+
+  eventDates(eventDate: Event['date'], workshopsDate: Event['workshopsDate']) {
```
`eventDate` is basically what you refer to as `conferenceDate` below. It does not include the workshops.

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+}
+
+function processDate(dates: any) {
```
Please create a `{start: date, end: date}` type and use it here as well as for `Event['date']` and `Event['workshopsDate']`.

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  let processedDate = MONTHS[startDate.getMonth()] + ' ' + startDate.getDate();
+
+  // If they are in different months add the string '- February 2' Making the final string January 31 - February 2
+  if (startDate.getMonth() !==  endDate.getMonth()) {
```
```suggestion
  if (startDate.getMonth() !== endDate.getMonth()) {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  // If they are in different months add the string '- February 2' Making the final string January 31 - February 2
+  if (startDate.getMonth() !==  endDate.getMonth()) {
+    processedDate = processedDate + ' - ' + MONTHS[endDate.getMonth()] + ' ' + endDate.getDate();
+  } else if (startDate.getTime() !== endDate.getTime()) {
```
```suggestion
  } else if (startDate.getDate() !== endDate.getDate()) {
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  }
+
+  return processedDate;
+}
```
Please use [template literals][1] in these two methods.

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

mary-poppins commented:
You can preview 08fc1e6 at https://pr36517-08fc1e6.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak can you please tell me why the test in events.service.spec.ts is failing I cannot figure it out.

mary-poppins commented:
You can preview bab76e9 at https://pr36517-bab76e9.ngbuilds.io/.

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,228 @@
+    "linkUrl": "https://www.angularconf.com.au/",
+    "name": "Angular Conf Australia",
+    "date": {
+      "start": "2018-07-22",
+      "end": "2018-07-22"
```
```suggestion
      "start": "2018-06-22",
      "end": "2018-06-22"
```

ajitsinghkaler replied:
Done

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,228 @@
+[
+  {
+    "location": "Oslo, Norway",
+    "tooltip": "ngVikings",
+    "linkUrl": "https://ngvikings.org/",
+    "name": "ngVikings",
+    "date": {
+      "start": "2020-05-25",
+      "end": "2020-05-26"
+    },  
+    "workshopsDate": {
+      "start": "2020-05-27",
+      "end": "2020-05-27"
+    }
```
Can you, please, sort the properties in the same order as the properties in the `Event` interface (for all entries)?

ajitsinghkaler replied:
I'll do this during clean up

ajitsinghkaler replied:
You want it in this order 
"location": "Oslo, Norway",
    "tooltip": "ngVikings",
    "linkUrl": "https://ngvikings.org/",
    "name": "ngVikings",
    "date": {
      "start": "2020-05-25",
      "end": "2020-05-26"
    },  
    "workshopsDate": {
      "start": "2020-05-27",
      "end": "2020-05-27"
    }

or alphabetical order

gkalpak replied:
In the order in which they appear on the `Event` interface: https://github.com/angular/angular/blob/a0af6605b211f3801d7abff1a4daf0bdaf7a829c/aio/src/app/custom-elements/events/events.component.ts#L27-L34

gkalpak commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,228 @@
+  },
+  {
+    "location": "Atlanta, Georgia",
+    "tooltip": "ngAtlanta",
```
Thinking about it again, I don't think there is any benefit in having a tooltip that is the same as the link text. Can you please remove all tooltips that are the same as the name? (Potentially in a separate commit.)

ajitsinghkaler replied:
Ok will do it in a separate commit

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.service.spec.ts:
```diff
@@ -0,0 +1,39 @@
+  afterEach(() => httpMock.verify());
+
+  it('should make a single connection to the server', () => {
+    eventsService.events.subscribe();
```
But multiple subscriptions will lead to multiple requests

ajitsinghkaler replied:
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  let processedDate = `${MONTHS[startDate.getMonth()]} ${startDate.getDate()}`;
+
+  // If they are in different months add the string '- February 2' Making the final string January 31 - February 2
+  if (startDate.getMonth()
+  !== endDate.getMonth()) {
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+  linkUrl: string;
+  tooltip?: string;
+
+  date: {
+    start: date;
+    end: date;
+  };
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    end: date;
+  };
+
+  workshopsDate?: {
+    start: date;
+    end: date;
+  };
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    });
+  }
+
+  geteventDates(eventDate: Duration, workshopsDate?: Duration) {
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    });
+  }
+
+  geteventDates(eventDate: Duration, workshopsDate?: Duration) {
```
No, it was just easy to get it from template

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+
+      if (eventDate.end > workshopsDate.start) {
+        // If workshop is in the end add (workshops) to the string after function processing
+        dateString = `${processDate(workshopsDate)} (workshops), ${processDate(eventDate)} (confrence)`;
```
done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+      // Check if the workshop is at the end
+      if (eventDate.start < workshopsDate.end) {
+        // If workshop is in the start add (confrence) to the string after function processing
+        dateString = `${processDate(eventDate)} (confrence), ${processDate(workshopsDate)} (workshops)`;
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    // Check if there is a workshop
+    if (workshopsDate) {
+      // Check if the workshop is at the start
+
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.service.spec.ts:
```diff
@@ -0,0 +1,39 @@
+      imports: [HttpClientTestingModule],
+      providers: [
+        EventsService,
+        { provide: Logger, useClass: TestLogger }
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.service.spec.ts:
```diff
@@ -0,0 +1,39 @@
+    const req = httpMock.expectOne({});
+    expect(req.request.url).toBe('generated/events.json');
+  });
+});
```
Done

ajitsinghkaler commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,228 @@
+    "location": "Salt Lake City, Utah",
+    "tooltip": "ng-conf",
+    "linkUrl": "https://ng-conf.org/",
+    "name": "ng-conf  ",
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+      if (eventDate.start < workshopsDate.end) {
+        // If workshop is in the start add (confrence) to the string after function processing
+        dateString = `${processDate(eventDate)} (confrence), ${processDate(workshopsDate)} (workshops)`;
+      }
```
Done

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+    } else {
+      dateString = processDate(eventDate);
+    }
+    // If no work shop date create confrence date string
```
Changed

mary-poppins commented:
You can preview 06ae9ac at https://pr36517-06ae9ac.ngbuilds.io/.

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,114 @@
+      this.upcomingEvents = events.filter(event => {
+        return new Date(event.date.end).getTime() >= Date.now() - DAY;
+      })
+      .sort((l: Event, r: Event) => l.date.start > r.date.start ? 1 : -1);
```
Done

mary-poppins commented:
You can preview 17fc486 at https://pr36517-17fc486.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak do you know why lint is failing

mary-poppins commented:
You can preview a0af660 at https://pr36517-a0af660.ngbuilds.io/.

gkalpak commented on aio/src/app/custom-elements/events/events.component.ts:
```diff
@@ -0,0 +1,112 @@
+        return new Date(event.date.end).getTime() < Date.now() - DAY;
+      })
+      .sort((l: Event, r: Event) => {
+        if (l.date.start < r.date.start) {
+          return 1;
+        } else if (l.date.start === r.date.start) {
+          return l.date.end < r.date.end ? 1 : -1;
+        } else { return -1; }
```
We can avoid the nested ternary here with:
 ```suggestion
        if (l.date.start < r.date.start) {
          return 1;
        } else if (l.date.start > r.date.start) {
          return -1;
        } else if (l.date.end < r.date.end) {
          return 1;
        } else {
          return -1;
        }
```

Which, in turn can be more simply written as:
```ts
        return (l.date.start < r.date.start) ||
            (l.date.start === r.date.start && l.date.end < r.date.end) ? 1 : -1;
```

With the equivalent for `upcomingEvents` being:
```ts
        return (l.date.start < r.date.start) ||
            (l.date.start === r.date.start && l.date.end < r.date.end) ? -1 : 1;
```

Which looks very similar to the logic for `areWorkshopsBeforeEvent` below:
```ts
const areWorkshopsBeforeEvent = (event.workshopsDate.start < event.date.start) ||
          (event.workshopsDate.start === event.date.start && event.workshopsDate.end < event.date.end);
```

##
So, where I am getting at with this is that we could extract this logic into a helper function and use it in all three places (making the code more DRY and readable):

```ts
function isBefore(duration1: Duration, duration2: Duration): boolean {
  return (duration1.start < duration2.start) ||
      (duration1.start === duration2.start && duration1.end < duration2.end);

// ...

this.pastEvents = events
    .filter(event => /* ... */)
    .sort((l: Event, r: Event) => isBefore(l.date, r.date) ? 1 : -1);

this.upcomingEvents = events
    .filter(event => /* ... */)
    .sort((l: Event, r: Event) => isBefore(l.date, r.date) ? -1 : 1);

// ...

const areWorkshopsBeforeEvent = isBefore(event.workshopsDate, event.date);

// Or directly (since it's readable enough):
dateString = isBefore(event.workshopsDate, event.date) ?
    `${workshopsDateString}, ${mainEventDateString}` :
    `${mainEventDateString}, ${workshopsDateString}`;
```

gkalpak reviewed (changes requested):
Getting closer :smiley: Thx, @ajitsinghkaler :+1:
Can you please squash the current fixup commits into the original one (since the PR has changed a lot since the beginning)? (Still add new changes in fixup commits - that helps a lot.)

mary-poppins commented:
You can preview 1560e4a at https://pr36517-1560e4a.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes now only sorting of keys and removing tooltip is left

mary-poppins commented:
You can preview ad30f96 at https://pr36517-ad30f96.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak can you please have a look

gkalpak commented:
@ajitsinghkaler, please make all the requested changes before I take another look.

ajitsinghkaler commented:
@gkalpak added all the changes should I sort right now and I added an extra test too that  that eventsService.events will emit an empty array? 

Can you please tell me what did I miss.

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.service.spec.ts:
```diff
@@ -0,0 +1,46 @@
+      [jasmine.any(Error)]
+    ]);
+    expect(mockLogger.output.error[0][0].message).toMatch(/^generated\/events\.json request failed:/);
+  });
```
added seprate test for this

ajitsinghkaler commented:
@gkalpak made the sorting changes and removed tooltip from the events.json can you please have a look

mary-poppins commented:
You can preview e2c1291 at https://pr36517-e2c1291.ngbuilds.io/.

gkalpak commented:
@ajitsinghkaler, is there any reason not to address all my comment (including sorting) before I take another look?
(It is a waste of time to take a look after a partially addressed review and try to figure out if missing changes are because you intended to make them later or you have missed something.)

gkalpak commented:
@ajitsinghkaler, can you please fix the CI failures? Ping me if you are not sure how to fix a specific failure.

mary-poppins commented:
You can preview 0ce08b0 at https://pr36517-0ce08b0.ngbuilds.io/.

mary-poppins commented:
You can preview cb32675 at https://pr36517-cb32675.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made ci green changed the order of properties and removed all extra tooltips

You can have a look now

ajitsinghkaler commented:
@gkalpak can you please have a look evrything is resolved

gkalpak reviewed (changes requested):
Thx, @ajitsinghkaler ✨

I've left some more comments. Also, please move the `change order of properties` commit into the previous commit. (They are supposed to be a single change.)

##
Note to self: Ensure there are testcases for all examples mentioned in https://github.com/angular/angular/pull/36517#discussion_r407496303.

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.service.spec.ts:
```diff
@@ -42,5 +43,15 @@ describe('EventsService', () => {
+  it('should return an empty array on a failed request for `events.json`', () => {
+    const request = httpMock.expectOne('generated/events.json');
+    request.error(new ErrorEvent('404'));
+    eventsService.events.subscribe(results => {
+      expect(results).toEqual([]);
+    });
+    expect().nothing();  // Prevent jasmine from complaining about no expectations.
```
@gkalpak why did you make events an observable in let events: Observable<Event[]> | null = null;

I don't think it should be an observable

gkalpak replied:
Yes, sorry. I meant `events: Event[] | null`. I'll fix the original code to avoid confusion.

ajitsinghkaler replied:
but in this case, the subscription will be async and expect(events).toEqual([]); will run before the subscription finishes and null will not be equal to an empty array

ajitsinghkaler replied:
But this will also not work because the events subscription is async and events will always be null in expect(events).toEqual([]);

ajitsinghkaler replied:
I've tried something please let me know

gkalpak replied:
LGTM, thx!

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -0,0 +1,184 @@
+});
+
+class TestEventService {
+  events = of(getEvents());
```
@gkalpak I did not get how to use this can you please help me. How to make the subject emit values is my real con fusion

gkalpak replied:
You can do something like:

```ts
const eventsService: TestEventsServive = injector.get(EventsService) as any;

// Use code that (directly or indirectly) subscribes to `eventsService.events`...

eventsService.events.emit([/* some tests events appropriate for this specific test */]);
```

Does that make sense?

We use this pattern i several places in the aio tests. You can see an example [here](https://github.com/angular/angular/blob/9166baf7090ae3172a9c5c3b233a6d191f0c4d5b/aio/src/app/custom-elements/search/file-not-found-search.component.spec.ts#L43).

ajitsinghkaler replied:
Yes resolved it

mary-poppins commented:
You can preview a82e618 at https://pr36517-a82e618.ngbuilds.io/.

mary-poppins commented:
You can preview 6f3c64c at https://pr36517-6f3c64c.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes please have a look

ajitsinghkaler commented:
@gkalpak can you please have a look

ajitsinghkaler commented:
@gkalpak can ou please review

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -0,0 +1,192 @@
+  describe('Getting event dates if no workshop date', () => {
+    it('should return only conference date', () => {
+      eventsService.events.next([testEvents[1]]);
+      const datestring = component.getEventDates(events[0]);
```
There doesn't seem to be a reason to emit events and grab them from `events`. Can you please just get rid of the `events` variable and test the `getEventDates()` method by passing events directly?

gkalpak commented:
Somehow I accidentally publiched the review comments before finishing the review, so here are some more comments:

- Please ensure there are tests cases for all the cases mentioned in https://github.com/angular/angular/pull/36517#discussion_r407496303
- Check if you still need to update the payload sizes and make the payload size changes on the same commit that made them necessary.

mary-poppins commented:
You can preview b4ad3c2 at https://pr36517-b4ad3c2.ngbuilds.io/.

mary-poppins commented:
You can preview eda62c7 at https://pr36517-eda62c7.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes

IgorMinar commented on aio/content/marketing/events.json:
```diff
@@ -0,0 +1,209 @@
+[
```
can you please add this new file into `.pullapprove` file under the group `docs-marketing` thanks!

ajitsinghkaler replied:
@IgorMinar it already is under `docs-marketing` via this rule https://github.com/angular/angular/blob/20346ff1924e768e0b3ff02f1539514263bf0e77/.pullapprove.yml#L775 So, no need to add it.

IgorMinar reviewed (dismissed):
overall this looks like a great change, but it still needs a bit more polish before it's merged.

I'll rely on @gkalpak to guide the technical review of this change. Outside of that I just want to ensure that the PR is rebased and the events.json is in sunc with events.html @ HEAD, and that we update the pullapprove config to account for this new file under the `docs-marketing` review group.

thanks!

Reviewed-for: docs-marketing

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
 
   describe('ngOnInit', () => {
     it('should have 2 upcoming events', () => {
```
This description does not help understand what is being tested (why 2). Please change it to something like:

```suggestion
    it('should have recognize upcoming events', () => {
```

Same for other tests below.

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
   describe('ngOnInit', () => {
     it('should have 2 upcoming events', () => {
-      eventsService.events.next(testEvents);
+      eventsService.events.next([{
+        'location': 'Oslo, Norway',
+        'tooltip': 'ngVikings',
+        'linkUrl': 'https://ngvikings.org/',
+        'name': 'ngVikings',
+        'date': {
+          'start': '2099-07-25',
+          'end': '2099-07-26'
+        },
+        'workshopsDate': {
+          'start': '2099-07-27',
+          'end': '2099-07-27'
+        }
+      },
+      {
+        'location': 'Salt Lake City, Utah',
+        'tooltip': 'ng-conf',
+        'linkUrl': 'https://ng-conf.org/',
+        'name': 'ng-conf  ',
+        'date': {
+          'start': '2099-06-20',
+          'end': '2099-06-20'
+        }
+      },
+      {
+        'location': 'Delhi, India',
+        'tooltip': 'ngIndia',
+        'linkUrl': 'https://www.ng-ind.com/',
+        'name': 'ngIndia',
+        'date': {
+          'start': '2020-02-29',
+          'end': '2020-02-29'
+        }
+      }]);
```
```suggestion
      eventsService.events.next([
        {
          'location': 'Oslo, Norway',
          'tooltip': 'ngVikings',
          'linkUrl': 'https://ngvikings.org/',
          'name': 'ngVikings',
          'date': {
            'start': '2099-07-25',
            'end': '2099-07-26',
          },
          'workshopsDate': {
            'start': '2099-07-27',
            'end': '2099-07-27',
          },
        },
        {
          'location': 'Salt Lake City, Utah',
          'tooltip': 'ng-conf',
          'linkUrl': 'https://ng-conf.org/',
          'name': 'ng-conf  ',
          'date': {
            'start': '2099-06-20',
            'end': '2099-06-20',
          },
        },
        {
          'location': 'Delhi, India',
          'tooltip': 'ngIndia',
          'linkUrl': 'https://www.ng-ind.com/',
          'name': 'ngIndia',
          'date': {
            'start': '2020-02-29',
            'end': '2020-02-29',
          },
        },
      ]);
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+          'start': '2018-07-27',
+          'end': '2018-07-27'
+        }
+      }]);
```
```suggestion
        }
      ]);
```

(And indent the rest of the items accordingly.)

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
-      eventsService.events.next([testEvents[1]]);
-      const datestring = component.getEventDates(events[0]);
+      const datestring = component.getEventDates({
+            'location': 'Salt Lake City, Utah',
+            'tooltip': 'ng-conf',
+            'linkUrl': 'https://ng-conf.org/',
+            'name': 'ng-conf  ',
+            'date': {
+              'start': '2020-06-20',
+              'end': '2020-06-20'
+            }
+          });
```
```suggestion
        'location': 'Salt Lake City, Utah',
        'tooltip': 'ng-conf',
        'linkUrl': 'https://ng-conf.org/',
        'name': 'ng-conf  ',
        'date': {
          'start': '2020-06-20',
          'end': '2020-06-20',
        },
      });
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+        'name': 'ReactiveConf',
+        'date': {
+          'start': '2019-10-30',
+          'end': '2019-11-01'
+        }
```
```suggestion
          'end': '2019-11-01',
        },
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+        'name': 'AngularConnect',
+        'date': {
+          'start': '2019-09-19',
+          'end': '2019-09-20'
+        }
```
```suggestion
          'end': '2019-09-20',
        },
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
-      const datestring = component.getEventDates(events[0]);
+    it('should return conference date with different days, workshopdate', () => {
+      const datestring = component.getEventDates({
+            'location': 'Oslo, Norway',
+            'tooltip': 'ngVikings',
+            'linkUrl': 'https://ngvikings.org/',
+            'name': 'ngVikings',
+            'date': {
+              'start': '2020-07-25',
+              'end': '2020-07-26'
+            },
+            'workshopsDate': {
+              'start': '2020-07-27',
+              'end': '2020-07-27'
+            }
+          });
```
```suggestion
        'location': 'Oslo, Norway',
        'tooltip': 'ngVikings',
        'linkUrl': 'https://ngvikings.org/',
        'name': 'ngVikings',
        'date': {
          'start': '2020-07-25',
          'end': '2020-07-26',
        },
        'workshopsDate': {
          'start': '2020-07-27',
          'end': '2020-07-27',
        },
      });
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+        'name': 'NG Rome MMXIX',
+        'date': {
+          'start': '2019-10-07',
+          'end': '2019-10-07'
+        },
+        'workshopsDate': {
+          'start': '2019-10-06',
+          'end': '2019-10-06'
+        }
```
```suggestion
          'end': '2019-10-07',
        },
        'workshopsDate': {
          'start': '2019-10-06',
          'end': '2019-10-06',
        },
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+        'name': 'NG-DE',
+        'date': {
+          'start': '2019-08-30',
+          'end': '2019-08-31'
+        },
+        'workshopsDate': {
+          'start': '2019-08-29',
+          'end': '2019-08-29'
+        }
```
```suggestion
          'end': '2019-08-31',
        },
        'workshopsDate': {
          'start': '2019-08-29',
          'end': '2019-08-29',
        },
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+        'name': 'ngVikings',
+        'date': {
+          'start': '2018-07-26',
+          'end': '2018-07-26'
+        },
+        'workshopsDate': {
+          'start': '2018-07-27',
+          'end': '2018-07-27'
+        }
```
```suggestion
          'end': '2018-07-26',
        },
        'workshopsDate': {
          'start': '2018-07-27',
          'end': '2018-07-27',
        },
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+
+    it('should return conference date and workshop date with different days and months, starting with confrence date', () => {
+      const datestring = component.getEventDates({
+            'location': 'Oslo, Norway',
+            'tooltip': 'ngVikings',
+            'linkUrl': 'https://ngvikings.org/',
+            'name': 'ngVikings',
+            'date': {
+              'start': '2020-07-30',
+              'end': '2020-07-31'
+            },
+            'workshopsDate': {
+              'start': '2020-08-01',
+              'end': '2020-08-02'
+            }
+          });
```
```suggestion
        'location': 'Oslo, Norway',
        'tooltip': 'ngVikings',
        'linkUrl': 'https://ngvikings.org/',
        'name': 'ngVikings',
        'date': {
          'start': '2020-07-30',
          'end': '2020-07-31',
        },
        'workshopsDate': {
          'start': '2020-08-01',
          'end': '2020-08-02',
        },
      });
```

gkalpak commented on aio/src/app/custom-elements/events/events.component.spec.ts:
```diff
@@ -31,162 +29,267 @@ describe('EventsComponent', () => {
+
+    it('should return conference date and workshop date with different days and months, starting with workshopdate', () => {
+      const datestring = component.getEventDates({
+            'location': 'Oslo, Norway',
+            'tooltip': 'ngVikings',
+            'linkUrl': 'https://ngvikings.org/',
+            'name': 'ngVikings',
+            'date': {
+              'start': '2020-08-01',
+              'end': '2020-08-02'
+            },
+            'workshopsDate': {
+              'start': '2020-07-30',
+              'end': '2020-07-31'
+            }
+          });
```
```suggestion
        'location': 'Oslo, Norway',
        'tooltip': 'ngVikings',
        'linkUrl': 'https://ngvikings.org/',
        'name': 'ngVikings',
        'date': {
          'start': '2020-08-01',
          'end': '2020-08-02',
        },
        'workshopsDate': {
          'start': '2020-07-30',
          'end': '2020-07-31',
        },
      });
```

ajitsinghkaler committed:
feat(docs-infra): created new widget for events page

Data in events page was hardcoded and it is manually moved in the table.

Created a new events widget which will automatically move past and upcoming
events from events.json (`aio/content/marketing/events.json`) file to the
relevant table in the events tab

ajitsinghkaler committed:
refactor(docs-infra): remove tooltip from links same as name

In the events.json file most of tooltips are same as name so there
were of no use, as they were providing no extra information. So,
removed them from the events.json file

gkalpak committed:
fixup! feat(docs-infra): created new widget for events page

gkalpak committed:
test(docs-infra): simplify `EventsComponent` tests and add more test cases

This commit simplifies the tests of `EventsComponent` (by introducing a
`createMockEvent()` helper and getting rid of the irrelevant `Event`
fields) and adds tests for some more usecases. It also makes the tests
more robust by using Jasmine's `Clock` to mock the current date.

mary-poppins commented:
You can preview 303c8f3 at https://pr36517-303c8f3.ngbuilds.io/.

mary-poppins commented:
You can preview 51e3def at https://pr36517-51e3def.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes please have a look

mary-poppins commented:
You can preview a84822c at https://pr36517-a84822c.ngbuilds.io/.

gkalpak commented:
@googlebot I consent.

mary-poppins commented:
You can preview 0b9c7ff at https://pr36517-0b9c7ff.ngbuilds.io/.

gkalpak commented:
BTW, I have confirmed that the events list is up-to-date with master (commit 20346ff1924e768e0b3ff02f1539514263bf0e77).

gkalpak commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -0,0 +1,43 @@
+  </thead>
+  <tbody >
+    <tr *ngFor="let event of upcomingEvents">
+      <td><a href="{{event.linkUrl}}" title="{{event.tooltip}}">{{event.name}}</a></td>
```
I just noticed that the first cell was a `<th>` in `events.html`. Is this change intentional? (If so why?)

ajitsinghkaler replied:
@gkalpak <th> is for headers <td> is for table data that's is why I changed it

gkalpak replied:
`<th>` does indeed define a cell as header of a group of table cells, but it is not necessary for that group to be a column (despite that being the most common usecase). There is nothing wrong with using `<th>` in a non-header row (i.e. a row inside `<tbody>`) that also contains `<td>` cells. See [<th> on MDN][1].

My main concern is that replacing `<th>` for `<td>` changes the styling of the tables, which is not something that this PR should do.

**Before**
![events-before](https://user-images.githubusercontent.com/8604205/84515130-bc467600-acd4-11ea-9778-44f8a9cff9b9.png)

**After**
![events-after](https://user-images.githubusercontent.com/8604205/84515147-c10b2a00-acd4-11ea-85e1-53f492218b25.png)

(I am not necessarily saying I prefer the current styling, but changing the style is not something that should happen on this PR.)

[1]: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/th

gkalpak commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -0,0 +1,43 @@
+<p>Where we'll be presenting:</p>
+<table class="is-full-width">
```
I think we should show a message (instead of an empty table) when there are no upcoming events.

mary-poppins commented:
You can preview 43e3a8a at https://pr36517-43e3a8a.ngbuilds.io/.

mary-poppins commented:
You can preview 5871d51 at https://pr36517-5871d51.ngbuilds.io/.

mary-poppins commented:
You can preview 09b5541 at https://pr36517-09b5541.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak made the changes. You can have a look.

Used some inline styling because I did not find the right place to put that css and creating a new file for text-align center seems too much

gkalpak reviewed (changes requested):
Thx, @ajitsinghkaler :+1:
Let's keep the `td --> th` change from the last fixup and merge this PR; then make additional changes to the `EventsComponent` in a another PR.

ajitsinghkaler commented on aio/src/app/custom-elements/events/events.component.html:
```diff
@@ -1,5 +1,5 @@
 <p>Where we'll be presenting:</p>
-<table class="is-full-width">
+<table class="is-full-width" *ngIf="upcomingEvents?.length; else noEvents">
```
yeah that works for me

ajitsinghkaler committed:
fixup! feat(docs-infra): created new widget for events page

ajitsinghkaler commented:
@gkalpak made the change

mary-poppins commented:
You can preview ea1c526 at https://pr36517-ea1c526.ngbuilds.io/.

gkalpak reviewed (approved):
Awesome! LGTM 💯

jelbourn reviewed (approved):
LGTM

Reviewed-for: size-tracking

ajitsinghkaler commented:
@IgorMinar can you please review this

mary-poppins commented:
You can preview fb33d8c at https://pr36517-fb33d8c.ngbuilds.io/.

ajitsinghkaler committed:
fixup! feat(docs-infra): created new widget for events page

mary-poppins commented:
You can preview 9e31903 at https://pr36517-9e31903.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak Is there a blocker for this

gkalpak commented:
No, it just needs to be reviewed afaict.

ajitsinghkaler commented:
Tha ks

jelbourn reviewed (approved):
LGTM

Reviewed-for: size-tracking

ajitsinghkaler commented:
@petebacondarwin can you approve this for size tracking

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

Although it is in conflict and needs a rebase

mary-poppins commented:
You can preview ae39c87 at https://pr36517-ae39c87.ngbuilds.io/.

mary-poppins commented:
You can preview 647476a at https://pr36517-647476a.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak pull approve has crashed. How to resolve it.

gkalpak commented:
No idea 😅 
@josephperrott might know.

gkalpak commented:
@ajitsinghkaler, could you rebase this and fix conflicts?

mary-poppins commented:
You can preview d73cfd2 at https://pr36517-d73cfd2.ngbuilds.io/.

ajitsinghkaler commented:
@gkalpak rebased it

gkalpak reviewed (approved):
Reviewed-for: global-docs-approvers

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers, size-tracking
